### PR TITLE
Relative objects in wasm

### PR DESCRIPTION
**Pull request by graydon**
This is a _much_ simpler approach to the "object read-access-control" problem covered in PRs #940 and #930 (and previously #132). The idea here is:

1. We split object references as before in two types differentiated by their low bit, but now they're "absolute" and "relative". Relative references indirect through a context-local table that just holds absolute references.
2. We only _really_ care about this issue when dealing with wasm contracts. Local-testing / native contracts have no isolation / complete system visibility anyways, there's no point really worrying about "who can see what".
3. Therefore the wasm VM argument-marshalling path is a natural point to do the relative-to-absolute translation. We translate relative references to absolute on the way in to the host, and absolute to relative on the way out. Wasm never gets to see any absolute references. If they forge one, we refuse it on the way in.

(There is a slight downside in the current implementation which is that if you return the same absolute reference multiple times we'll create new relative references on each return. If this is annoying we can use a hashtable or something instead of a vector. But the vector has both performance and conceptual simplicity on its side, and in any case the user _shouldn't_ care about specific object references, and I expect most contracts do not run long enough or return the same object references repeatedly enough for this to matter either way)

### CODE REVIEW

**Comment by graydon** [comment]
@dmkozh I believe this is ready to go now, and is as minimal as I can make it. I added a test and localized the changes to the host (and actually _just_ the VM frame type). At this point more of the diff is just legibility change (I changed the VM frame type from a tuple to a struct) than anything else.

**Review by jayz22 — APPROVED** [review]

**Review by dmkozh — APPROVED** [review]

File: soroban-env-host/src/test/hostile.rs [hunk 1]
```diff
@@ -116,3 +116,34 @@ fn hostile_objs_traps() -> Result<(), HostError> {
...
+    host.set_diagnostic_level(crate::DiagnosticLevel::Debug)?;
+    host.with_budget(|b| b.reset_default())?;
+    host.with_budget(|b| b.reset_unlimited_cpu())?;
+
+    let private_vec = host.vec_new_from_slice(&[1u32.into(), 2u32.into()])?;
+
+    let payload = private_vec.to_val().get_payload();
+    let lo = payload as u32;
+    let hi = (payload >> 32) as u32;
+    let args = host.vec_new_from_slice(&[lo.into(), hi.into()])?;
```
[the remark this answers is not in the document]
Comment (graydon) [reply soroban-env-host/src/test/hostile.rs:134]: I'll add a few more variants like that.

[the remark this answers is not in the document]
Comment (graydon) [reply soroban-env-host/src/test/hostile.rs]: (This actually found lots of interesting new edge cases. Hooray for more tests!)
